src/main/java/org/candlepin/model/RhsmApiCompatCurator.java

```diff
@@ -103,4 +116,421 @@ public List<ConsumerEntitlementCount> getConsumerEntitlementCounts(Collection<St
+                "ON glu.guest_id = guest.guest_id AND glu.last_updated = guest.updated " +
+                "WHERE hypervisor.owner_id = :owner_id";
+        }
+
+        List<Object[]> rows = entityManager.createNativeQuery(sql)
+            .setParameter("owner_id", ownerId)
+            .getResultList();
+
+        Map<String, HypervisorGuestMapping> result = new HashMap<>();
+        for (Object[] row : rows) {
+            String hypervisorId = (String) row[0];
+            String hypervisorUuid = (String) row[1];
+            String hypervisorName = (String) row[2];
+            String guestId = (String) row[3];
+            HypervisorGuestMapping mapping = new HypervisorGuestMapping(
+                hypervisorId, hypervisorName, hypervisorUuid, guestId);
+
+            // Since we know there will be few consumer records, we can preload all mappings.
+            // This avoids looping through all consumer feeds for each lookup and improves performance.
+            List<String> guestIdVariants = Util.getPossibleUuids(guestId);
+            for (String variant : guestIdVariants) {
+                result.put(variant.toLowerCase(), mapping);
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Retrieves a mapping from consumer ID to a list of {@link ConsumerInstalledProduct} for all
+     * installed products belonging to consumers matching the specified owner and filter criteria.
+     *
+     * @param ownerId      the ID of the owner whose consumers' products are to be fetched; must not be null
+     * @param afterId      (optional) only consumers with an ID lexicographically greater than this value
+     *                     will be included
+     * @param afterUuid    (optional) only consumers with a UUID lexicographically greater than this value
+     *                     will be included
+     * @param afterCheckin (optional) only consumers with a last check-in timestamp after this value will
+     *                     be included
+     * @param page         the page number to retrieve
+     * @param perPage      the number of consumers to return per page
+     * @return a map where each key is a consumer ID and each value is a list of installed products
+     * for that consumer
+     */
+    private Map<String, Set<ConsumerFeedInstalledProduct>> getInstalledProductsByConsumer(String ownerId,
+        String afterId, String afterUuid, OffsetDateTime afterCheckin, int page, int perPage) {
+        String cte = buildConsumersSubquery(afterId, afterUuid, afterCheckin);
+        String sql =
+            "SELECT ip.consumer_id, ip.product_id, ip.product_name, ip.product_version " +
+                "FROM cp_installed_products ip " +
+                "JOIN (" + cte + ") cids ON cids.id = ip.consumer_id";
+
+        Query q = entityManager.createNativeQuery(sql);
+        q.setParameter("owner_id", ownerId);
+        if (afterId != null) {
+            q.setParameter("after_id", afterId);
```

Comment:
These are scary -- we're setting parameters blindly that do not have any way of guaranteeing their presence. 

An example of an alternate approach is in the Product/ContentQueryBuilder, which passes a map of arguments around instead of having disconnected logic setting values.

Reply:
That's a pretty good idea, I didn't think of that. I added a generic BuiltSql record that we can use and wrote some documentation with an example.